### ISSUE TITLE
Zero downtime ES deployments

### ISSUE THREAD

**Issue by mrdavidlaing** [logsearch/logsearch-boshrelease#51]
@dpb587 asks

> How can we adjust our bosh deployments to upgrade/restart elasticsearch nodes one at a time (serial = true), but then wait for green cluster before continuing with the next node?

**Comment by mrdavidlaing** [comment]
@wdneto - do you have any ideas here?

**Comment by aristotelesneto** [comment]
This can be accomplished by combining a couple of things together.

1) Override the global "update:" on the job you only want 1 node to be restarted at a time on the deployment manifest, and set "max_in_flight: 1". This ensures only one node is ever worked on at a time.

2) Setup a drain script for the job.  The way drain scripts work, is that before bosh sends a shutdown command it runs the drain script, meaning that it won't prevent BOSH from continuing to the next node per se, but instead it will prevent the next node from going down before knowing the cluster is ok (ie. the script runs on the next node, not on the node you just upgraded).

Simply doing option 2 should suffice, but I think it's probably best to do both.

The drain script isn't really documented, but [here is where BOSH handles it](https://github.com/cloudfoundry/bosh/blob/master/bosh_agent/lib/bosh_agent/message/drain.rb). 
A couple of major quirks:
- The sample script has to be named "bin/drain" (destination file on the job spec)
- If you're collocating multiple jobs together, the drain script will only run for the first job specified in the manifest (bug, reported but not yet fixed) 
- The drain script expects only an integer as output (on stdout, any other text will cause it to error), as well as an exit code of 0. 
  - Exiting with anything other than 0 means you will need to ssh into the machine to fix the reason why (ie. theres no command you can send to bosh from the outside to tell it to ignore the exit code of the drain script).
  - If the output of the script is a negative integer (eg: -N) , bosh wait N seconds before sending another drain command, however, when it does the arguments of the script changes, which I found very annoying.

Here is [a basic sample](https://gist.github.com/wdneto/b6f64c805652ddfb60ca) I used for removing an OSD from a Ceph cluster when a node is deleted.

The useful available options are job_shutdown (when the node is being deleted, useful for removing itself from a cluster, triggering permanent evacuation) and job_changed (job is being updated, but should still remain in the cluster).

**Comment by aristotelesneto** [comment]
[Here is another sample drain script](https://github.com/cloudfoundry/cf-release/blob/master/jobs/dea_next/templates/deterministic_drain.rb), from the DEA.

**Comment by mrdavidlaing** [comment]
@wdneto - Thanks - these features open up lots of possibilities.

A question:

Consider a deployment spread over multiple AZs, with 1 ES job (with 2 instances) / AZ

```
jobs:
- name: elasticsearch_z1
  instances: 2
  update:
    serial: true
    max_in_flight: 1
...
- name: elasticsearch_z2
  instances: 2
  update:
    serial: true
    max_in_flight: 1
```

Will this update in the following serial order? `elasticsearch_z1/0` => `elasticsearch_z1/1` => `elasticsearch_z2/0` => `elasticsearch_z2/1` ?

**Comment by dpb587** [comment]
@wdneto, I agree it's a very helpful solution - thanks!

@mrdavidlaing, in your example, I was assuming that `serial`/`max_in_flight` was job-specific, so `z1/0` and `z2/0` would both go offline for updates concurrently. It seems like BOSH would require a separate `update.group` sort of setting in order to correlate multiple jobs together into a single update workflow.

**Comment by aristotelesneto** [comment]
@dpb587 it's better to think about the serial option in the global context, as opposed to job specific. The equivalent of 'serial' for a job specific is essentially max_in_flight: 1 and canaries: 1.

If your default option is serial: false, and you have a manifest like so:

```
jobs:
- name: job1
- name: job2
- name: job3
...
- name: elasticsearch_z1
  update:
    serial: true
    max_in_flight: 1
...
- name: job4
- name: job5
- name: elasticsearch_z2
  instances: 2
  update:
    serial: true
    max_in_flight: 1
```

BOSH will start job1/0, job2/0, job3/0 in parallel, each with their respective (or the global) canaries/max_in_flight options.

Then it spins up elasticsearch_z1/0 by itself (provided canaries:1), then z1/1.

Then it will spin up job4/0, job5/0 in parallel.

And finally, elasticsearch_z2/0, then z2/1.

So the progression for the example @mrdavidlaing mentioned would be z1/0, z1/1, then proceed to z2/0 and z2/1.

**Comment by dpb587** [comment]
Oh, interesting; thanks for explaining that.

**Comment by mrdavidlaing** [comment]
To talk specifics; I propose we change our biggest logsearch cluster's jobs look like this:

```
update:
  serial: true          # default to serial updates to keep overall cluster online
  max_in_flight: 1  # only update 1 instance in a job group at a time

jobs:
- name: ingestor
  instances: 1
  update:
    serial: false
- name: queue
  instances: 1
  update:
    serial: false

- name: api
  instances: 1

- name: log_parser-1a
  instances: 1
  update:
    serial: false
- name: log_parser-1b
  instances: 1
  update:
    serial: false
- name: log_parser-1c
  instances: 1
  update:
    serial: false

- name: elasticsearch_eu-west-1a
  instances: 2
- name: elasticsearch_eu-west-1b
  instances: 2
- name: elasticsearch_eu-west-1c
  instances: 2
```

Which should lead to a deployment pattern of:
- ingestor, queue
- api
- log_parser-1a/b/c
- elasticsearch_eu-west-1a/0
- elasticsearch_eu-west-1a/1
- elasticsearch_eu-west-1b/0
- elasticsearch_eu-west-1b/1
- elasticsearch_eu-west-1c/0
- elasticsearch_eu-west-1c/1

Have I interpreted how this works correctly?
Is this the deploy behaviour we want?

**Comment by aristotelesneto** [comment]
Hrmm, cant say I tried the anti-pattern :-)

That would be mostly equivalent to:

```
update:
  serial: false
  max_in_flight: 1  # only update 1 instance in a job group at a time

jobs:
- name: ingestor
  instances: 1
- name: queue
  instances: 1

- name: api
  instances: 1
  update:
    serial: true

- name: log_parser-1a
  instances: 1
- name: log_parser-1b
  instances: 1
- name: log_parser-1c
  instances: 1

- name: elasticsearch_eu-west-1a
  instances: 2
- name: elasticsearch_eu-west-1b
  instances: 2
  update:
    serial: true
- name: elasticsearch_eu-west-1c
  instances: 2
```

The pattern would then be:
- ingestor, queue
- api
- log_parser-1a/b/c, elasticsearch_eu-west-1a/0
  - elasticsearch_eu-west-1a/1
- elasticsearch_eu-west-1b/0
  - elasticsearch_eu-west-1b/1
- elasticsearch_eu-west-1c/0
  - elasticsearch_eu-west-1c/1

**Comment by dpb587** [comment]
I missed referencing this issue in this commit:

> https://github.com/logsearch/logsearch-boshrelease/commit/0dc44f9998faebf8f9f2df58d0587842dc85a8a9

I've added a drain (and [slightly hacky] undrain) script and I'm excited about how it will dramatically help deployments:

> https://github.com/logsearch/logsearch-boshrelease/blob/develop/jobs/elasticsearch/templates/bin/drain.erb
> https://github.com/logsearch/logsearch-boshrelease/blob/develop/jobs/elasticsearch/templates/bin/undrain.erb

It significantly improves recovery times after node restarts (it disables allocations, avoiding the automatic rebalancing which blocked up recovery and caused later rebalancing). It also takes care of scaling down the cluster if a job is being destroyed (it forces moving all the shards off the node and waits until that finishes).

After quite a bit of testing locally, I deployed it to meta and went through a few deploys successfully.

This should have been a PR, but I accidentally committed an earlier version today, and then was too excited after I got the rest working to want to revert it.

@wdneto, @mrdavidlaing - if you have a moment, do you have any feedback about the drain/undrain script?

**Comment by mrdavidlaing** [comment]
@dpb587 This looks very cool; although I'll admit I'm finding it a little difficult to visualise what is going on.

Any chance of a diagram?

**Comment by aristotelesneto** [comment]
:+1: This looks great! 

I'm not entirely familiar with the specific ES settings you're changing via the API, but I assume they mean the following. Forgive me if they already have the intended outcome, and this is simply my ignorance.

```
{"transient":{"cluster.routing.allocation.enable":"primaries"}}
```

Remove me from selection for allocation of index other than primary ones?

Wouldn't you want to cause all indexes to move, regardless of whether they're primary or not? Better yet, is it possible for ES to maintain the indexes on a node, and only provide it with the updates while it was down? The latter option would prevent significant amounts of wasted I/O, since we know this is a transient state, and the node is coming back eventually (depends on our tolerance for having the cluster in a degraded state while the node is down).

```
'{"drained":1}' 'localhost:9200/.bosh/drain/<%= name %>--<%= index %>' 
```

Didn't realise ES allowed for "random path" API calls... Pretty cool if it does.

```
'{"transient":{"cluster.routing.allocation.exclude._name":"<%= name %>/<%= index %>"}}'
```

Remove me from selection for allocation of index. I'm assuming that this is all that's required for shutting down a node and deleting it has no bearing on the cluster itself (ie. the cluster will no longer know anything about the node that's just been deleted)

Other than that I've made some comments on the code itself, just to try and help a bit with readability :-)

Just for my own benefit, how are these API calls different from the Shutdown API / index allocation method described in:
http://www.elasticsearch.org/guide/en/elasticsearch/reference/current/setup-upgrade.html#rolling-upgrades

**Comment by dpb587** [comment]
Here's a general overview to both of your comments which I think will further clarify things...
- indices have 1+ shards to segment all the data
- indices have 0+ replicas - extra copies of the data
- shards are either primaries or replicas - primaries are the "authority" responsible for the data; replicas are copies
- elasticsearch cluster has 1+ nodes - in our case elasticsearch node = elasticsearch process
- elasticsearch nodes are responsible for a subset of shards
- elasticsearch cluster manages balancing the shards/replicas according to available disk space and custom attributes (e.g. across multiple AZs)
- elasticsearch color status: green = all shards available; yellow = all primary shards available; red = primary shards missing

When nodes (and their shards) go offline during restarts, elasticsearch will immediately try to allocate the missing shards to other nodes (from another shard from an online node, if available) to ensure data availability. This obviously creates additional disk and network I/O to copy those multi-GB shards. There are three side effects to this. First, when the restarting node comes back online and sees its shard has been allocated to another node, it will delete its local shard because it's no longer needed in the cluster. Second, this causes a longer delay in cluster recovery because recovering the shard across the network is obviously slower than recovering from its original disk. Third, once all shards are back online elasticsearch will realize the restarted node has more disk space than the other nodes, so it'll rebalance the shard back (typically and effectively moving the shard back to its original node - a waste of disk/network I/O and time).

> The drain script changes `cluster.routing.allocation.enable` to `primaries` to avoid all this. By setting allocation to `primaries`, elasticsearch is allowed to shift responsibility for any primary shards from the restarting node onto another node (thus keeping the shard and index online). Because we write data synchronously, this is safe since primaries and replicas should be consistent at whatever point a node decides to shut down. By setting allocation to `primaries`, elasticsearch is not allowed to assign missing replicas to any other node (avoiding the thrashing around of all that data, unnecessarily).
> 
> Once the node is back online, the undrain script changes the allocation settting back to normal. Since the node is online, it can accurately report back the shards it has, so elasticsearch allocates the shards back to it.
> 
> The `.bosh/drain/<%= name %>--<%= index %>` key is used to avoid overriding changes which may have been done manually.

When a node is getting shut down completely, all its shards will obviously be lost. When it is gone, elasticsearch will start trying to recover all those lost shards onto online nodes using any available replicas. This creates disk/network load on an irregular number of nodes (shard source and shard target) with multiple nodes possibly trying to be both the source and target of a recovery (further increasing the disk I/O).

> The drain script changes `cluster.routing.allocation.exclude._name` to the name of the node being shut down. This setting tells elasticsearch that that node by that name should not be allocated any shards, so elasticsearch starts the process of moving any shards it has onto other nodes. This restricts the replication shard source to a single node and, due to regular disk space balancing, it will be unlikely that a single node is the replication target for more than one shard.
> 
> The drain script then blocks until the node is responsible for zero shards. This process can take quite a while (unlike the prior, simple restart case where nodes are reloading data from local disk) because it probably needs to transfer a decent amount of data over the network.

About the drain script, it really only cares about the `job_status` argument. It seems to be invoked first and once with the actual task (`changed`/`unchanged` vs `shutdown`) with subsequent invocations trying to check the `status`. The drain script may be invoked before the cluster is ready for new cluster operations (i.e. the cluster is still recovering from node 2's restart, but bosh is ready to start node 3's restart), so the first invocation it's storing the job task. Then, once the cluster is green, internally it makes sure it runs the task-specific commands once which may result in the cluster becoming yellow. Then, we go back to the normal case of waiting for green before allowing BOSH to manipulate the node.

---

In terms of the `_shutdown` API, we can't use that call directly because it primarily just ends the process (and monit will automatically restart it). The `cluster.routing.allocation.enable` setting though is what's being used in drain, although the guide is a bit different than what I'd been experiencing. I'll re-evaluate it with their suggestion of using `none` instead of the current `primaries` - I might have misinterpreted things when I was learning.

I'm going to revert these changes and move this discussion to a PR where I should have made it originally.

**Comment by dpb587** [comment]
Let's move further discussion to the new PR #56 dedicated to this topic and related code changes, but I'll keep this open until the PR is merged and this issue resolved.
